diff --git a/drivers/ata/libata-core.c b/drivers/ata/libata-core.c
index c5ba15a..5df566a 100644
--- a/drivers/ata/libata-core.c
+++ b/drivers/ata/libata-core.c
@@ -69,6 +69,7 @@
 #include <linux/ratelimit.h>
 #include <linux/pm_runtime.h>
 #include <linux/platform_device.h>
+#include <trace/events/power.h>
 
 #include "libata.h"
 #include "libata-transport.h"
@@ -3289,6 +3290,7 @@ static int ata_dev_set_mode(struct ata_device *dev)
 		     ata_mode_string(ata_xfer_mode2mask(dev->xfer_mode)),
 		     dev_err_whine);
 
+	trace_suspend_resume(TPS("ATA"), ap->print_id, false);
 	return 0;
 
  fail:
