diff --git a/drivers/ata/libata-core.c b/drivers/ata/libata-core.c
index 18d97d5..6e3b8b6 100644
--- a/drivers/ata/libata-core.c
+++ b/drivers/ata/libata-core.c
@@ -68,6 +68,7 @@
 #include <linux/ratelimit.h>
 #include <linux/pm_runtime.h>
 #include <linux/platform_device.h>
+#include <trace/events/power.h>
 
 #include "libata.h"
 #include "libata-transport.h"
@@ -3288,6 +3289,7 @@ static int ata_dev_set_mode(struct ata_device *dev)
 		     ata_mode_string(ata_xfer_mode2mask(dev->xfer_mode)),
 		     dev_err_whine);
 
+	trace_suspend_resume(TPS("ATA"), ap->print_id, false);
 	return 0;
 
  fail:
