diff --git a/drivers/base/power/main.c b/drivers/base/power/main.c
index 86d5e4f..d784339 100644
--- a/drivers/base/power/main.c
+++ b/drivers/base/power/main.c
@@ -880,6 +880,7 @@ static void device_complete(struct device *dev, pm_message_t state)
 {
 	void (*callback)(struct device *) = NULL;
 	char *info = NULL;
+	ktime_t calltime;
 
 	if (dev->power.syscore)
 		return;
@@ -907,7 +908,9 @@ static void device_complete(struct device *dev, pm_message_t state)
 
 	if (callback) {
 		pm_dev_dbg(dev, state, info);
+		calltime = initcall_debug_start(dev);
 		callback(dev);
+		initcall_debug_report(dev, calltime, 0, state, info);
 	}
 
 	device_unlock(dev);
