diff --git a/kernel/power/suspend.c b/kernel/power/suspend.c
index bef86d1..c6a3afa 100644
--- a/kernel/power/suspend.c
+++ b/kernel/power/suspend.c
@@ -265,7 +265,9 @@ int suspend_devices_and_enter(suspend_state_t state)
 			goto Close;
 	}
 	suspend_console();
+#ifndef CONFIG_PM_DEBUG
 	ftrace_stop();
+#endif
 	suspend_test_start();
 	error = dpm_suspend_start(PMSG_SUSPEND);
 	if (error) {
@@ -285,7 +287,9 @@ int suspend_devices_and_enter(suspend_state_t state)
 	suspend_test_start();
 	dpm_resume_end(PMSG_RESUME);
 	suspend_test_finish("resume devices");
+#ifndef CONFIG_PM_DEBUG
 	ftrace_start();
+#endif
 	resume_console();
  Close:
 	if (need_suspend_ops(state) && suspend_ops->end)
