diff --git a/drivers/base/power/runtime.c b/drivers/base/power/runtime.c
index 3148b10..d694d1a 100644
--- a/drivers/base/power/runtime.c
+++ b/drivers/base/power/runtime.c
@@ -11,6 +11,7 @@
 #include <linux/export.h>
 #include <linux/pm_runtime.h>
 #include <trace/events/rpm.h>
+#include <linux/suspend.h>
 #include "power.h"
 
 static int rpm_resume(struct device *dev, int rpmflags);
@@ -480,6 +481,33 @@ static int rpm_suspend(struct device *dev, int rpmflags)
 	goto out;
 }
 
+static ktime_t rpm_resume_debug_start(struct device *dev)
+{
+	ktime_t calltime = ktime_set(0, 0);
+
+	if (pm_print_times_enabled) {
+		pr_info("calling  %s+ @ %i, parent: %s, rpm_resume\n",
+			dev_name(dev), task_pid_nr(current),
+			dev->parent ? dev_name(dev->parent) : "none");
+		calltime = ktime_get();
+	}
+
+	return calltime;
+}
+
+static void rpm_resume_debug_report(struct device *dev, ktime_t calltime,
+				  int error)
+{
+	ktime_t delta, rettime;
+
+	if (pm_print_times_enabled) {
+		rettime = ktime_get();
+		delta = ktime_sub(rettime, calltime);
+		pr_info("call %s+ returned %d after %Ld usecs, rpm_resume\n", dev_name(dev),
+			error, (unsigned long long)ktime_to_ns(delta) >> 10);
+	}
+}
+
 /**
  * rpm_resume - Carry out runtime resume of given device.
  * @dev: Device to resume.
@@ -500,6 +528,7 @@ static int rpm_suspend(struct device *dev, int rpmflags)
 static int rpm_resume(struct device *dev, int rpmflags)
 	__releases(&dev->power.lock) __acquires(&dev->power.lock)
 {
+	ktime_t calltime;
 	int (*callback)(struct device *);
 	struct device *parent = NULL;
 	int retval = 0;
@@ -652,7 +681,10 @@ static int rpm_resume(struct device *dev, int rpmflags)
 	if (!callback && dev->driver && dev->driver->pm)
 		callback = dev->driver->pm->runtime_resume;
 
+	calltime = rpm_resume_debug_start(dev);
 	retval = rpm_callback(callback, dev);
+	rpm_resume_debug_start(dev, calltime, retval);
+
 	if (retval) {
 		__update_runtime_status(dev, RPM_SUSPENDED);
 		pm_runtime_cancel_pending(dev);
